core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -794,18 +890,8 @@ private[spark] class SparkSubmit extends Logging {
       logInfo("\n")
     }
 
-    val loader =
-      if (sparkConf.get(DRIVER_USER_CLASS_PATH_FIRST)) {
-        new ChildFirstURLClassLoader(new Array[URL](0),
-          Thread.currentThread.getContextClassLoader)
-      } else {
-        new MutableURLClassLoader(new Array[URL](0),
-          Thread.currentThread.getContextClassLoader)
-      }
-    Thread.currentThread.setContextClassLoader(loader)
-
     for (jar <- childClasspath) {
-      addJarToClasspath(jar, loader)
+      addJarToClasspath(jar, getSubmitClassLoader(sparkConf))
```

Comment:
This looks broken. `getSubmitClassLoader` will create a new class loader for each jar file. So every time you call that method, you're tossing out all the work you did previously.

Reply:
Ok, `getSubmitClassLoader` will just chain class loaders, so you will not actually lose the previous state, but you'll end up with an unnecessary long class loader hierarchy, instead of having just one class loader specific for the app.

Reply:
what is the correct idiom here?  call `val loader = getSubmitClassLoader(sparkConf)` once and then `addJarToClasspath(jar, loader)` in the loop?

Reply:
That's what the previous code does, so I'd start there.

Reply:
I just need that before runMain, so I need a proper way to know that this has been called once, I guess it could be a singleton.